File: python/worker.py
#!/usr/bin/env python
import logging
logging.basicConfig()

import argparse
import json
import os

import pika

import sys
here_dir = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, os.path.join(here_dir, 'user_code'))


class Worker(object):

    QUEUE_NAME = 'tasks'
    QUEUE_PORT = 5672

    def __init__(self, queue_host, queue_port=None, queue_name=None):
        self.queue_host = queue_host
        self.queue_port = queue_port or self.QUEUE_PORT
        self.queue_name = queue_name or self.QUEUE_NAME
        self.connect()

    def connect(self):
        """Establish a connection with the task queue."""

        connection = pika.BlockingConnection(
            pika.ConnectionParameters(host=self.queue_host,
                                      port=self.queue_port))
        self.channel = connection.channel()
        self.channel.queue_declare(queue=self.queue_name, durable=True)

    def run(self):
        self.channel.basic_qos(prefetch_count=1)
        self.channel.basic_consume(self.callback, queue=self.queue_name)
        self.channel.start_consuming()

    @staticmethod
    def callback(channel, method, properties, body):
        print('Received:', body)
        try:
            result = self.process(body)
        except Exception as exc:
            result = {'error': exc.message}
        channel.basic_publish(exchange='',
                              routing_key=properties.reply_to,
                              body=json.dumps(result))
        channel.basic_ack(delivery_tag=method.delivery_tag)

    def process(self, body):
        import main
        metadata = json.load(open('metadata.json'))
        result = main.process(json.loads(body))
        return result


def parse_args():
    parser = argparse.ArgumentParser(description='A Python worker',
                                     epilog='--*--')
    parser.add_argument('--queue-host', metavar='HOST',
                        help='host where RabbitMQ is running')
    parser.add_argument('--queue-port', metavar='PORT',
                        type=int, default=5672,
                        help='port where RabbitMQ is running')
    parser.add_argument('-i', '--interactive', action='store_true',
                        help='run interactive console')
    return parser.parse_args()


def main():
    args = parse_args()
    if args.interactive:
        os.execlp('ipython', 'ipython')
    worker = Worker(args.queue_host, args.queue_port)
    print('Worker starting')
    worker.run()


if __name__ == '__main__':
    main()

File: apim/containers/javascript/worker.js
#!/usr/bin/env node

var amqplib = require('amqplib');

function consume(host, port, queue) {
    url = 'amqp://' + host + ':' + port;
    amqplib.connect(url).then(function(conn) {
        return conn.createChannel().then(function(ch) {
            var ok = ch.assertQueue(queue, {durable: true});
            ok = ok.then(function(_qok) {
                return ch.consume(queue, on_consume, {noAck: true});
            });
            return ok;

            function on_consume(msg) {
                var lines = [];
                var responder = function(data) {
                    ch.publish('', msg.properties.replyTo,
                               new Buffer(data));
                };
                console.log = function(data) {
                    if (data === '---' || data === 'END') {
                        responder(lines.join(' '));
                        lines = [];
                    } else {
                        lines.push(data);
                    }
                    if (data === 'END') {
                        responder(data);
                    }
                };
                try {
                    var main = require('./user_code/main.js');
                    main.process(JSON.parse(msg.content.toString()));
                    responder(JSON.stringify({'time_in_main': 0}));
                }
                catch (err) {
                    responder(JSON.stringify({error: err.message}));
                    responder('END');
                    responder(JSON.stringify({'time_in_main': 0}));
                }
            };

        });
    }).then(null, function (msg) {
        console.error('*** WORKER ERROR');
        console.error(msg);
    });
}

function worker() {
    var host = process.argv[3];
    var port = process.argv[5];
    var queue = process.argv[7];
    console.error('*** WORKER STARTED');
    consume(host, port, queue);
}

worker();
